### DISCUSSION TITLE
Unable to compile

### DISCUSSION THREAD

**Discussion by ghost** [google/or-tools#3479] [Build and configuration issues]
Hi,

I'm trying to compile a program in Ubuntu (20.04) using VS2019 in Windows via WSL but I'm getting the following error message:
_/usr/include/ortools/linear_solver/linear_solver.h(149,10): error : absl/base/port.h: No such file or directory_

I've followed the official guide (https://developers.google.com/optimization/install/cpp/source_linux) and its seems the examples run just fine.

Am I doing something wrong?

Thanks!

**Comment by Mizux** [comment]
what do you means by "a program" a custom one ? then can you share the CMakeLists.txt you did use ? the command line used ? any releavant trace ?

seems we miss a dependenciy on absl/base/port.h which is part of `absl::core_headers`
ref: https://github.com/abseil/abseil-cpp/blob/40ede06edcc02fa1762bb0e445cef833f5856fba/absl/base/CMakeLists.txt#L121-L137

which is not part of https://github.com/google/or-tools/blob/b3599cf67099b0bc66ef1ecd7bd0376c0c4d5ecb/cmake/deps.cmake#L35-L55 try to add it here ;)

ps: Will try to fix asap tomorrow, 00:26am here time to sleep...

**Comment by ghost** [comment]
Yes, it's a project I developed first in Windows and need to compile it now for linux. 

I just followed the guide:
```sh
sudo apt-get -y install git wget pkg-config build-essential cmake autoconf libtool zlib1g-dev lsb-release
git clone https://github.com/google/or-tools
make third_party
make cc
make test_cc
make cpp
```

The ortools directory was actually created in /usr/include/ but no sign of the absl directory...

Thank you for your help!

**Comment by Mizux** [comment]
take a look at `install_make/` we perform a "local install"....

ps: `make cpp` is sufficient this will call `make third_party` and `*_cc` rules are just alias to `*_cpp` and are deprecated.
pps: you could also directly build using the cmake based build
see: https://github.com/google/or-tools/tree/main/cmake#cmake-options

ppps: Makefile just call the cmake based build now https://github.com/google/or-tools/blob/b3599cf67099b0bc66ef1ecd7bd0376c0c4d5ecb/makefiles/Makefile.cpp.mk#L75-L82

**Comment by ghost** [comment]
Well, it seems that building the library from the source is beyond my very limited skills.

Today I've tried with both the stable and the main source code but the directory /usr/include/ortools is not created, and hence I cannot link my code when compiling. Although running ``make test_cpp_linear_solver_samples`` works fine.

After downloading the source code, I simply ran ``make cpp``. What am I doing wrong? Could it be a problem with my Ubuntu (WSL)?

ps: Yesterday I said that the directory /ortools/ had been created but maybe it was me who copied it to there...

**Reply by Mizux** [reply]
which release of Ubuntu does WSL provide you ? you may take a look at 
https://github.com/google/or-tools/blob/6a1a0553fe23100913b96070cadd964f250b94ac/tools/docker/images/ubuntu-22.04.Dockerfile#L10-L13

when running `make cpp` what did you see ?

**Reply by ghost** [reply]
I'm using Ubuntu 20.04.

I've attatched part of the output when running ``make cpp``.
[make cpp.txt](https://github.com/google/or-tools/files/9685038/make.cpp.txt)

If I try to build my own project with VS2019 I get:
_error : ortools/linear_solver/linear_solver.h: No such file or directory_

I think that yesterday I may have copied the folder ortools into /usr/include/, which changed the error message to:
_/usr/include/ortools/linear_solver/linear_solver.h(149,10): error : absl/base/port.h: No such file or directory_

Also, after copying folder absl into /usr/include/, started getting errors such as:
_linear_solver.pb.h: No such file or directory_

**Reply by Mizux** [reply]
You should take a look at this rusty project:
https://github.com/or-tools/cmake_or-tools/tree/main/FetchContent
here: https://github.com/or-tools/cmake_or-tools/blob/a29b5907ac3e49a6de2ca4bd5bf6fd16cf58632f/FetchContent/CMakeLists.txt#L32-L42
seems to still work:
https://github.com/or-tools/cmake_or-tools/actions/runs/2715829485
